### PR TITLE
Ln/embed tiktok

### PR DESCRIPTION
**Pull request by lazynina**

### CODE REVIEW

File: src/app/backend-api.service.ts [hunk 1]
```diff
@@ -1259,6 +1261,13 @@ export class BackendApiService {
     });
   }
 
+  GetFullTikTokURL(endpoint: string, UpdaterPublicKeyBase58Check: string, TikTokShortVideoID: string): Observable<any> {
+    console.log("over here");
```
Comment (maebeam) [inline src/app/backend-api.service.ts:1265]: rm

File: src/app/backend-api.service.ts [hunk 2]
```diff
@@ -1259,6 +1261,13 @@ export class BackendApiService {
...
+    return this.jwtPost(endpoint, BackendRoutes.RoutePathGetFullTikTokURL, UpdaterPublicKeyBase58Check, {
```
Comment (maebeam) [inline src/app/backend-api.service.ts:1266]: don't see a reason why this needs to be jwt post?

File: src/app/feed/feed-create-post/feed-create-post.component.html [hunk 3]
```diff
@@ -50,7 +50,11 @@
   </div>
 </div>
 <div class="flex-fill px-15px br-10px" [ngClass]="{ 'd-none': !showEmbedVideoURL }">
-  <input class="br-3px" type="url" [(ngModel)]="embedVideoURL" placeholder="Enter a url to a youtube or vimeo video" />
+  <input class="br-3px"
+         type="url"
+         [(ngModel)]="embedVideoURL"
+         (ngModelChange)="setEmbedVideoURL()"
+         placeholder="Enter a url to a youtube or vimeo video" />
```
Comment (maebeam) [inline src/app/feed/feed-create-post/feed-create-post.component.html:57]: or tiktok?

File: src/lib/services/video-url-parser-service/video-url-parser-service.ts [hunk 4]
```diff
@@ -93,10 +165,27 @@ export class VideoUrlParserService {
...
     return false;
   }
+
+  static getEmbedHeight(link: string): number {
+    if (this.isValidVimeoEmbedURL(link) || this.isValidYoutubeEmbedURL(link)) {
+      return 315;
+    }
+    if (this.isValidTiktokEmbedURL(link)) {
+      return 700;
+    }
```
Comment (maebeam) [inline src/lib/services/video-url-parser-service/video-url-parser-service.ts:188]: maybe just `} else { return 315; }`?

File: src/app/feed/feed-post/feed-post.component.ts [hunk 5]
```diff
@@ -392,8 +394,20 @@ export class FeedPostComponent implements OnInit {
...
   }
 
-  getEmbedVideoURLForPostContent(): string {
-    return VideoUrlParserService.getEmbedVideoURL(this.postContent.PostExtraData["EmbedVideoURL"]);
+  getEmbedVideoURLForPostContent(): any {
+    return this.constructedEmbedVideoURL;
+  }
+
+  async setEmbedVideoURLForPostContent(): Promise<void> {
+    this.constructedEmbedVideoURL = await VideoUrlParserService.getEmbedVideoURL(
```
Comment (maebeam) [inline src/app/feed/feed-post/feed-post.component.ts:402]: Let's not use the async / await pattern here. Return a promise for `VideoUrlParserService.getEmbedVideoURL` and then call `VideoUrlParserService.getEmbedVideoURL().subscribe(res => this.constructEmbedVideoUrl = res`

File: src/app/feed/feed-post/feed-post.component.ts [hunk 6]
```diff
@@ -99,15 +99,17 @@ export class FeedPostComponent implements OnInit {
   hidingPost = false;
   quotedContent: any;
   _blocked: boolean;
+  constructedEmbedVideoURL: any;
 
-  ngOnInit() {
+  async ngOnInit() {
```
Comment (maebeam) [inline src/app/feed/feed-post/feed-post.component.ts:104]: drop all asyncs / awaits

File: src/lib/services/video-url-parser-service/video-url-parser-service.ts [hunk 7]
```diff
@@ -35,7 +37,61 @@ export class VideoUrlParserService {
...
-  static getEmbedVideoURL(embedVideoURL: string): string {
+  static async tiktokParser(
+    backendApi: BackendApiService,
+    globalVars: GlobalVarsService,
+    url: string
+  ): Promise<string | boolean> {
+    let fullTikTokURL = url;
+    try {
+      const tiktokURL = new URL(url);
+      if (tiktokURL.hostname === "vm.tiktok.com") {
```
Comment (maebeam) [inline src/lib/services/video-url-parser-service/video-url-parser-service.ts:48]: If we're doing the validation serverside I'd skip this

File: src/lib/services/video-url-parser-service/video-url-parser-service.ts [hunk 8]
```diff
@@ -35,7 +37,61 @@ export class VideoUrlParserService {
...
+        const regExp = /^.*(vm\.tiktok\.com\/)([A-Za-z0-9]{6,12}).*/;
+        const match = url.match(regExp);
+        if (match && match[2]) {
+          fullTikTokURL = await backendApi
+            .GetFullTikTokURL(globalVars.localNode, globalVars.loggedInUser.PublicKeyBase58Check, match[2])
```
Comment (maebeam) [inline src/lib/services/video-url-parser-service/video-url-parser-service.ts:53]: just return `backendApi. GetFullTikTokURL`

File: src/app/backend-api.service.ts [hunk 9]
```diff
@@ -1259,6 +1261,13 @@ export class BackendApiService {
...
+      TikTokShortVideoID,
+    });
```
Comment (maebeam) [inline src/app/backend-api.service.ts:1268]: I'd suggest putting the implementation detail of parsing the response here. Look in this file for examples of calling `map()` on a Promise and mapping the response to `FullTikTokURL`

**Review by maebeam — CHANGES_REQUESTED** [review]

File: src/lib/services/video-url-parser-service/video-url-parser-service.ts (hunk 7 above)
Comment (lazynina) [reply src/lib/services/video-url-parser-service/video-url-parser-service.ts:49]: saves us a trip to the server, let's keep it.

File: src/app/backend-api.service.ts (hunk 9 above)
Comment (lazynina) [reply src/app/backend-api.service.ts:1267]: I missed this one. let me know if you think it's still cleaner to move over here.

Comment (lazynina) [reply src/app/backend-api.service.ts]: I'm in favor of keeping the logic that extracts the video ID from the URL in the video url parser service and not calling it in the backend api service.  Adding it there would introduce a circular dependency

**Review by maebeam — APPROVED** [review]

**Pull request merged**